File: content/manual/v8.0/source/includes/aggregation/aggregation-examples/group-and-total/full-files/group-total.c
#include <stdio.h>
#include <bson/bson.h>
#include <mongoc/mongoc.h>

int main(void)
{
    mongoc_init();

    char *uri = "<connection string>";
    mongoc_client_t *client = mongoc_client_new(uri);
    // start-insert-orders
    mongoc_collection_t *orders = mongoc_client_get_collection(client, "agg_tutorials_db", "orders");

    {
        bson_t *filter = bson_new();
        bson_error_t error;
        if (!mongoc_collection_delete_many(orders, filter, NULL, NULL, &error))
        {
            fprintf(stderr, "Delete error: %s\n", error.message);
        }
        bson_destroy(filter);
    }

    {
        size_t num_docs = 9;
        bson_t *docs[num_docs];

        docs[0] = BCON_NEW(
            "customer_id", BCON_UTF8("elise_smith@myemail.com"),
            "orderdate", BCON_DATE_TIME(1590825352000UL), // 2020-05-30T08:35:52Z
            "value", BCON_INT32(231));

        docs[1] = BCON_NEW(
            "customer_id", BCON_UTF8("elise_smith@myemail.com"),
            "orderdate", BCON_DATE_TIME(1578904327000UL), // 2020-01-13T09:32:07Z
            "value", BCON_INT32(99));

        docs[2] = BCON_NEW(
            "customer_id", BCON_UTF8("oranieri@warmmail.com"),
            "orderdate", BCON_DATE_TIME(1577865937000UL), // 2020-01-01T08:25:37Z
            "value", BCON_INT32(63));

        docs[3] = BCON_NEW(
            "customer_id", BCON_UTF8("tj@wheresmyemail.com"),
            "orderdate", BCON_DATE_TIME(1559061212000UL), // 2019-05-28T19:13:32Z
            "value", BCON_INT32(2));

        docs[4] = BCON_NEW(
            "customer_id", BCON_UTF8("tj@wheresmyemail.com"),
            "orderdate", BCON_DATE_TIME(1606171013000UL), // 2020-11-23T22:56:53Z
            "value", BCON_INT32(187));

        docs[5] = BCON_NEW(
            "customer_id", BCON_UTF8("tj@wheresmyemail.com"),
            "orderdate", BCON_DATE_TIME(1597793088000UL), // 2020-08-18T23:04:48Z
            "value", BCON_INT32(4));

        docs[6] = BCON_NEW(
            "customer_id", BCON_UTF8("elise_smith@myemail.com"),
            "orderdate", BCON_DATE_TIME(1608963346000UL), // 2020-12-26T08:55:46Z
            "value", BCON_INT32(4));

        docs[7] = BCON_NEW(
            "customer_id", BCON_UTF8("tj@wheresmyemail.com"),
            "orderdate", BCON_DATE_TIME(1614496172000UL), // 2021-02-28T07:49:32Z
            "value", BCON_INT32(1024));

        docs[8] = BCON_NEW(
            "customer_id", BCON_UTF8("elise_smith@myemail.com"),
            "orderdate", BCON_DATE_TIME(1601722184000UL), // 2020-10-03T13:49:44Z
            "value", BCON_INT32(102));

        bson_error_t error;
        if (!mongoc_collection_insert_many(orders, (const bson_t **)docs, num_docs, NULL, NULL, &error))
        {
            fprintf(stderr, "Insert error: %s\n", error.message);
        }

        for (int i = 0; i < num_docs; i++)
        {
            bson_destroy(docs[i]);
        }
    }
    // end-insert-orders

    {
        const bson_t *doc;
        bson_t *pipeline = BCON_NEW(
            "pipeline", "[",
            // start-match
            "{", "$match", "{",
            "orderdate", "{",
            "$gte", BCON_DATE_TIME(1577836800000UL), // Represents 2020-01-01T00:00:00Z
            "$lt", BCON_DATE_TIME(1609459200000UL),  // Represents 2021-01-01T00:00:00Z
            "}",
            "}", "}",
            // end-match
            // start-sort1
            "{", "$sort", "{", "orderdate", BCON_INT32(1), "}", "}",
            // end-sort1
            // start-group
            "{", "$group", "{",
            "_id", BCON_UTF8("$customer_id"),
            "first_purchase_date", "{", "$first", BCON_UTF8("$orderdate"), "}",
            "total_value", "{", "$sum", BCON_UTF8("$value"), "}",
            "total_orders", "{", "$sum", BCON_INT32(1), "}",
            "orders", "{", "$push", "{",
            "orderdate", BCON_UTF8("$orderdate"),
            "value", BCON_UTF8("$value"),
            "}", "}",
            "}", "}",
            // end-group
            // start-sort2
            "{", "$sort", "{", "first_purchase_date", BCON_INT32(1), "}", "}",
            // end-sort2
            // start-set
            "{", "$set", "{", "customer_id", BCON_UTF8("$_id"), "}", "}",
            // end-set
            // start-unset
            "{", "$unset", "[", BCON_UTF8("_id"), "]", "}",
            // end-unset
            "]");

        // start-run-agg
        mongoc_cursor_t *results =
            mongoc_collection_aggregate(orders, MONGOC_QUERY_NONE, pipeline, NULL, NULL);

        bson_destroy(pipeline);
        // end-run-agg

        while (mongoc_cursor_next(results, &doc))
        {
            char *str = bson_as_canonical_extended_json(doc, NULL);
            printf("%s\n", str);
            bson_free(str);
        }
        bson_error_t error;
        if (mongoc_cursor_error(results, &error))
        {
            fprintf(stderr, "Aggregation error: %s\n", error.message);
        }

        mongoc_cursor_destroy(results);
    }

    // start-cleanup
    mongoc_collection_destroy(orders);
    // end-cleanup
    mongoc_client_destroy(client);
    mongoc_cleanup();

    return EXIT_SUCCESS;
}

File: content/manual/v8.0/source/includes/aggregation/aggregation-examples/group-and-total/full-files/group_total.go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

// start-structs
type Order struct {
	CustomerID string        `bson:"customer_id,omitempty"`
	OrderDate  bson.DateTime `bson:"orderdate"`
	Value      int           `bson:"value"`
}

// end-structs

func main() {
	const uri = "<connection string>"

	client, err := mongo.Connect(options.Client().ApplyURI(uri))
	if err != nil {
		log.Fatal(err)
	}

	defer func() {
		if err = client.Disconnect(context.TODO()); err != nil {
			log.Fatal(err)
		}
	}()

	aggDB := client.Database("agg_tutorials_db")

	// start-insert-orders
	orders := aggDB.Collection("orders")
	orders.DeleteMany(context.TODO(), bson.D{})

	_, err = orders.InsertMany(context.TODO(), []interface{}{
		Order{
			CustomerID: "elise_smith@myemail.com",
			OrderDate:  bson.NewDateTimeFromTime(time.Date(2020, 5, 30, 8, 35, 53, 0, time.UTC)),
			Value:      231,
		},
		Order{
			CustomerID: "elise_smith@myemail.com",
			OrderDate:  bson.NewDateTimeFromTime(time.Date(2020, 1, 13, 9, 32, 7, 0, time.UTC)),
			Value:      99,
		},
		Order{
			CustomerID: "oranieri@warmmail.com",
			OrderDate:  bson.NewDateTimeFromTime(time.Date(2020, 1, 01, 8, 25, 37, 0, time.UTC)),
			Value:      63,
		},
		Order{
			CustomerID: "tj@wheresmyemail.com",
			OrderDate:  bson.NewDateTimeFromTime(time.Date(2019, 5, 28, 19, 13, 32, 0, time.UTC)),
			Value:      2,
		},
		Order{
			CustomerID: "tj@wheresmyemail.com",
			OrderDate:  bson.NewDateTimeFromTime(time.Date(2020, 11, 23, 22, 56, 53, 0, time.UTC)),
			Value:      187,
		},
		Order{
			CustomerID: "tj@wheresmyemail.com",
			OrderDate:  bson.NewDateTimeFromTime(time.Date(2020, 8, 18, 23, 4, 48, 0, time.UTC)),
			Value:      4,
		},
		Order{
			CustomerID: "elise_smith@myemail.com",
			OrderDate:  bson.NewDateTimeFromTime(time.Date(2020, 12, 26, 8, 55, 46, 0, time.UTC)),
			Value:      4,
		},
		Order{
			CustomerID: "tj@wheresmyemail.com",
			OrderDate:  bson.NewDateTimeFromTime(time.Date(2021, 2, 29, 7, 49, 32, 0, time.UTC)),
			Value:      1024,
		},
		Order{
			CustomerID: "elise_smith@myemail.com",
			OrderDate:  bson.NewDateTimeFromTime(time.Date(2020, 10, 3, 13, 49, 44, 0, time.UTC)),
			Value:      102,
		},
	})
	if err != nil {
		log.Fatal(err)
	}
	// end-insert-orders

	// start-match
	matchStage := bson.D{{Key: "$match", Value: bson.D{
		{Key: "orderdate", Value: bson.D{
			{Key: "$gte", Value: time.Date(2020, 1, 1, 0, 0, 0, 0, time.UTC)},
			{Key: "$lt", Value: time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC)},
		}},
	}}}
	// end-match

	// start-sort1
	sortStage1 := bson.D{{Key: "$sort", Value: bson.D{
		{Key: "orderdate", Value: 1},
	}}}
	// end-sort1

	// start-group
	groupStage := bson.D{{Key: "$group", Value: bson.D{
		{Key: "_id", Value: "$customer_id"},
		{Key: "first_purchase_date", Value: bson.D{{Key: "$first", Value: "$orderdate"}}},
		{Key: "total_value", Value: bson.D{{Key: "$sum", Value: "$value"}}},
		{Key: "total_orders", Value: bson.D{{Key: "$sum", Value: 1}}},
		{Key: "orders", Value: bson.D{{Key: "$push", Value: bson.D{
			{Key: "orderdate", Value: "$orderdate"},
			{Key: "value", Value: "$value"},
		}}}},
	}}}
	// end-group

	// start-set
	setStage := bson.D{{Key: "$set", Value: bson.D{
		{Key: "customer_id", Value: "$_id"},
	}}}
	// end-set

	// start-sort2
	sortStage2 := bson.D{{Key: "$sort", Value: bson.D{
		{Key: "first_purchase_date", Value: 1},
	}}}
	// end-sort2

	// start-unset
	unsetStage := bson.D{{Key: "$unset", Value: bson.A{"_id"}}}
	// end-unset

	// start-run-agg
	pipeline := mongo.Pipeline{matchStage, sortStage1, groupStage, setStage, sortStage2, unsetStage}
	cursor, err := orders.Aggregate(context.TODO(), pipeline)
	// end-run-agg

	if err != nil {
		log.Fatal(err)
	}

	defer func() {
		if err := cursor.Close(context.TODO()); err != nil {
			log.Fatalf("failed to close cursor: %v", err)
		}
	}()

	var results []bson.D
	if err = cursor.All(context.TODO(), &results); err != nil {
		log.Fatalf("failed to decode results: %v", err)
	}
	for _, result := range results {
		res, _ := bson.MarshalExtJSON(result, false, false)
		fmt.Println(string(res))
	}
}
